### PR TITLE
Add hardest code blocks ever to the abstract-ship

### PR DESCRIPTION
**Pull request by bocharsky-bw**
Resolve `05-abstract-ship` in #1

### CODE REVIEW

File: knpu/episode3/abstract-ship.md [hunk 1]
```diff
@@ -4,15 +4,22 @@ There is one more thing that is special about the Rebel Ships. Since, they're
 the good guys we're going to give them some extra Jedi power. 
 
 Inside of `Ship` we have a `jediFactor` which is a value that is set from the
-database and a `getJediFactor` function. In the `BattleManager` this is used to 
-figure out if some super awesome Jedi powers are used during the battle. 
+database and a `getJediFactor` function:
```
Comment (weaverryan) [inline knpu/episode3/abstract-ship.md:7]: there were missing from the original transcription, but I prefer to always have functions with the `()`: `getJediFactor()`

File: knpu/episode3/abstract-ship.md [hunk 2]
```diff
@@ -4,15 +4,22 @@ There is one more thing that is special about the Rebel Ships. Since, they're
...
 have at least some Jedi Power, sometimes there's a lot and sometimes it's lower,
 depending on what side of the galaxy they woke up on that day. So, instead of making 
 this a dynamic value that we set in the datbase let's create a `public function getJediFactor()` 
-that returns the `rand` function with levels between 10 and 30. Setting it up like 
-this overrides the function in the `Ship` parent class. 
+that returns the `rand` function with levels between 10 and 30:
+
+[[[ code('410ffbb658') ]]]
+
+Setting it up likethis overrides the function in the `Ship` parent class.
```
Comment (weaverryan) [inline knpu/episode3/abstract-ship.md:22]: whoops :) `like this`

File: knpu/episode3/abstract-ship.md [hunk 3]
```diff
@@ -23,9 +30,15 @@ Over in PhpStorm, when we look at this function now, `Ship` has a Jedi Factor pr
...
-the `isFunctional` method. In `RebelShip` it's always true, but in `Ship` it reads from an `underRepair`
-property, and again that's just not needed in `RebelShip`. The point being, `Ship` comes with
-extra stuff that we are inheriting but not using in `RebelShip`. 
+the `isFunctional` method. In `RebelShip` it's always true:
+
+[[[ code('4bab20796d') ]]]
+
+but in `Ship` it reads from an `underRepair` property, and again that's just not needed in `RebelShip`:
+
+[[[ code('d89d34a269') ]]]
```
Comment (weaverryan) [inline knpu/episode3/abstract-ship.md:39]: I took out the highlighting of `__construct()` where `underRepair` is set - I don't think that detail is important enough.

File: knpu/episode3/abstract-ship.md [hunk 4]
```diff
@@ -36,48 +49,89 @@ This is a way of changing the class heirachy so that each class has only what it
...
-into `AbstractShip`. I know this looks like where we just were, but trust me we're going somewhere with
-this.
+Let's start this! Create a new PHP Class called `AbstractShip`:
+
+[[[ code('7e9d688d11') ]]]
+
+Since it is the most abstract idea of a ship in our project. To start, I'm going to copy everything
+out of the `Ship` class and paste it into `AbstractShip`:
+
+[[[ code('3d459c1edc') ]]]
```
Comment (weaverryan) [inline knpu/episode3/abstract-ship.md:59]: This one is hard: if we show the entire class, it's way too big of a code block. So, I selected a few properties and a few methods - and hopefully with the `...` between the sections and the text describing what we're doing, the user will get the idea.

File: knpu/episode3/abstract-ship.md [hunk 5]
```diff
@@ -36,48 +49,89 @@ This is a way of changing the class heirachy so that each class has only what it
...
+
+I know this looks like where we just were, but trust me we're going somewhere with this.
+
+Now, let's write `Ship extends AbstractShip`:
+
+[[[ code('90643e4c44') ]]]
+
+and do the same thing in `RebelShip` changing it from `Ship` to `AbstractShip`:
```
Comment (weaverryan) [inline knpu/episode3/abstract-ship.md:67]: I usually try to always capitalize the first letter - unless this line is definitely not a full sentence, for example:

``` md
Update Ship

<code>

and JediShip

<code>
```

File: knpu/episode3/abstract-ship.md [hunk 6]
```diff
@@ -36,48 +49,89 @@ This is a way of changing the class heirachy so that each class has only what it
...
-And `RebelShip` no longer inherits the `jediFactor` property and anything that works with it. Now each file
-is simpler, and only has the code that it actually needs. Back to the browser to test that everything still
-works. Oh look an error! `Call to undefined method RebelShip::setJediFactor() on ShipLoader line 55`. Let's
-check that out. 
+Now let's move `setJediFactor` from `AsbtractShip` into `Ship`:
+
+[[[ code('7c0010172b') ]]]
+
+and that should do it! Now, `Ship` still has all the functioanlity that it had before,
+it extends `AbstractShip`, and only contains it's unique code. And `RebelShip` no longer inherits
```
Comment (weaverryan) [inline knpu/episode3/abstract-ship.md:105]: our bad: `it's` => `its`

**Comment by weaverryan** [comment]
:+1: nice work! I made the changes I suggested - most of my comments are more conversational, so we can do these the same way. If you have any ideas or disagree with any suggestions, that's good :) - let me know

File: knpu/episode3/abstract-ship.md (hunk 5 above)
Comment (bocharsky-bw) [inline knpu/episode3/abstract-ship.md:67]: Oh, I did it many times.. I was considering it as a sentence continuation.

OK, I'll capitalize first letters in the future for similar cases.

File: knpu/episode3/abstract-ship.md (hunk 1 above)
Comment (bocharsky-bw) [inline knpu/episode3/abstract-ship.md:7]: Yes, I was missing it too. I always try to add `()` to the method/function name

**Pull request merged**